File: backend/app/Models/Poll.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Str;

class Poll extends Model
{
    protected $fillable = ['slug', 'question', 'options', 'active', 'closes_at'];

    protected $casts = [
        'options' => 'array',
        'active' => 'boolean',
        'closes_at' => 'datetime',
    ];

    protected static function booted(): void
    {
        static::saving(function (Poll $p) {
            if (empty($p->slug)) {
                $p->slug = Str::slug($p->question);
            }
            // Normalize options — ensure each has id/label/votes.
            if (is_array($p->options)) {
                $p->options = array_map(function ($o, $i) {
                    return [
                        'id' => $o['id'] ?? Str::slug($o['label'] ?? "opt-$i"),
                        'label' => $o['label'] ?? '',
                        'votes' => (int) ($o['votes'] ?? 0),
                    ];
                }, $p->options, array_keys($p->options));
            }
        });
    }

    public function votes(): HasMany
    {
        return $this->hasMany(PollVote::class);
    }

    public function scopeActive(Builder $q): Builder
    {
        return $q->where('active', true);
    }
}

File: app/lib/models/poll.dart
class PollOption {
  final String id;
  final String label;
  final int votes;

  PollOption({required this.id, required this.label, required this.votes});

  factory PollOption.fromJson(Map<String, dynamic> j) => PollOption(
        id: j['id'] as String,
        label: j['label'] as String,
        votes: (j['votes'] as int?) ?? 0,
      );
}

class Poll {
  final String slug;
  final String question;
  final List<PollOption> options;
  final bool active;
  final DateTime? closesAt;
  final int totalVotes;

  Poll({
    required this.slug,
    required this.question,
    required this.options,
    required this.active,
    this.closesAt,
    required this.totalVotes,
  });

  factory Poll.fromJson(Map<String, dynamic> j) => Poll(
        slug: j['slug'] as String,
        question: j['question'] as String,
        options: ((j['options'] as List<dynamic>?) ?? [])
            .map((o) => PollOption.fromJson(o as Map<String, dynamic>))
            .toList(),
        active: (j['active'] as bool?) ?? true,
        closesAt: j['closesAt'] != null
            ? DateTime.tryParse(j['closesAt'] as String)
            : null,
        totalVotes: (j['totalVotes'] as int?) ?? 0,
      );
}
